src/BuiltInTools/dotnet-watch/HotReloadDotNetWatcher.cs

```diff
@@ -186,13 +192,41 @@ public async Task WatchAsync(DotNetWatchContext context, CancellationToken cance
             }
         }
 
-        private static void ConfigureExecutable(DotNetWatchContext context, ProcessSpec processSpec)
+        private void ConfigureExecutable(DotNetWatchContext context, ProcessSpec processSpec)
         {
             var project = context.FileSet.Project;
-            processSpec.Executable = project.RunCommand;
-            if (!string.IsNullOrEmpty(project.RunArguments))
+
+            // For webapps, we can support changes to Program.Main and Startup by restarting the host (aka hot restart).
+            // First determine if we can perform hot restarts.
+            if (!SuppressHotRestart &&
+                ProjectGraph.EntryPointNodes.FirstOrDefault().ProjectInstance is { } projectInstance &&
+                !string.IsNullOrEmpty(projectInstance.GetPropertyValue("UsingMicrosoftNETSdkWeb")))
             {
-                processSpec.EscapedArguments = project.RunArguments;
+                // When performing a hot restart, we have to run a different assembly than the project's output. This
+                // is similar to how the inside dll works for dotnet-ef. Infer metadata from the project to allow the
+                // alternate entry point to run.
+                var projectDll = projectInstance.GetPropertyValue("TargetPath");
```

Comment:
Do we need to be defensive here at all? Are these always set?

Reply:
It'd be very unusual for it to not be set: https://github.com/dotnet/msbuild/blob/500b60ac4b827fbaa38c784dc5c6df8cc254dc33/src/Tasks/Microsoft.Common.CurrentVersion.targets#L324. There's other msbuild targets (I know for e.g. Razor used to rely on it) that would fail if this is not configured correctly.